### ISSUE TITLE
HLS segment name collision when using the lowest latency buffer option

### ISSUE THREAD

**Issue by zerodytrash** [owncast/owncast#1778]
hi,
When I start a stream with "Medium latency", it works without problems, but when I activate the "lowest" latency option, the stream runs for several hours without problems and then starts to buffer every few seconds. My first thought was that the HLS segments are not written fast enough so that the browser has to request them again. But that is not the problem. After debugging I noticed the following:

At the lowest latency buffer option, the segment length is one second. The segment filename contains a Unix timestamp (seconds since January 1st, 1970). Therefore, a unique filename can be generated every second. The ffmpeg parameter `hls_time=1` specifies that a video segment is one second long. But the parameter does not guarantee that the elapsed time between two written segments is always exactly one second. And that seems to be the problem here.

I first looked at the `m3u8` playlist in the browser developer console. There are two segments with the same name. This happens again and again and the player buffers every few seconds.

```
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:30:55.000-0800
stream-hYTBAjLng1646944255.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:30:56.000-0800
stream-hYTBAjLng1646944255.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:30:57.000-0800
stream-hYTBAjLng1646944256.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:30:58.000-0800
stream-hYTBAjLng1646944257.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:30:59.000-0800
stream-hYTBAjLng1646944258.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:31:00.000-0800
stream-hYTBAjLng1646944260.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:31:01.000-0800
stream-hYTBAjLng1646944261.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:31:02.000-0800
stream-hYTBAjLng1646944261.ts
#EXTINF:1.000000,
#EXT-X-PROGRAM-DATE-TIME:2022-03-10T12:31:03.000-0800
stream-hYTBAjLng1646944262.ts
```
(`stream-hYTBAjLng1646944255.ts` and `stream-hYTBAjLng1646944261.ts` are present twice)

The Owncast debug output: 
```
DEBU[2022-03-10T12:31:02-08:00] [hls muxer @ 0x564197d5c7c0] Duplicated segment filename detected: stream-hYTBAjLng1646944261.ts
```

After that, I printed out every received `PUT` request with milliseconds:

```
12:30:52.738 Receive PUT /0/stream-hYTBAjLng1646944252.ts
12:30:53.801 Receive PUT /0/stream.m3u8
12:30:53.801 Receive PUT /0/stream-hYTBAjLng1646944253.ts
12:30:55.213 Receive PUT /0/stream.m3u8
12:30:55.213 Receive PUT /0/stream-hYTBAjLng1646944255.ts
12:30:55.737 Receive PUT /0/stream.m3u8
12:30:55.738 Receive PUT /0/stream-hYTBAjLng1646944255.ts
12:30:56.728 Receive PUT /0/stream.m3u8
12:30:56.728 Receive PUT /0/stream-hYTBAjLng1646944256.ts
12:30:57.730 Receive PUT /0/stream.m3u8
12:30:57.730 Receive PUT /0/stream-hYTBAjLng1646944257.ts
12:30:58.737 Receive PUT /0/stream.m3u8
12:30:58.737 Receive PUT /0/stream-hYTBAjLng1646944258.ts
12:31:00.471 Receive PUT /0/stream.m3u8
12:31:00.471 Receive PUT /0/stream-hYTBAjLng1646944260.ts
12:31:01.123 Receive PUT /0/stream.m3u8
12:31:01.123 Receive PUT /0/stream-hYTBAjLng1646944261.ts
12:31:01.741 Receive PUT /0/stream.m3u8
12:31:01.742 Receive PUT /0/stream-hYTBAjLng1646944261.ts
12:31:02.741 Receive PUT /0/stream.m3u8
12:31:02.741 Receive PUT /0/stream-hYTBAjLng1646944262.ts
12:31:03.750 Receive PUT /0/stream.m3u8
```

You can see a segment named `stream-hYTBAjLng1646944255.ts` was received at `12:30:55.213` and again at `12:30:55.738` (525 millis diff)
But the video length of both segments with the same name is exactly one second long. The time difference to the segment which was written before (`stream-hYTBAjLng1646944253.ts`) is greater than one second. Probably due to a delay in the RTMP connection or during encoding the output process was just delayed. And then suddenly it runs faster again and ffmpeg writes the next segment with less latency.

**In summary, this means**: We cannot assume that the interval between writing the segments is exactly one second. If the interval is sometimes only 900 milliseconds (due to fluctuations in RTMP or encoding latency), the naming of the segments does not work reliably.

Would it be possible to include milliseconds or a random string in the filename to prevent collisions?

**Comment by zerodytrash** [comment]
Owncast v0.0.12-dev
Ubuntu 20.04
ffmpeg 4.2.4-1ubuntu0.1

**Comment by gabek** [comment]
This is a great catch, thanks so much for looking into it!  It totally makes sense. Using the timestamp in the filename was just an identifier and isn't used for anything, so maybe it's best just to remove it so this can't happen.  I'll fix this up shortly!

**Comment by zerodytrash** [comment]
At the moment I use the filename timestamp for a VOD service. I query all HLS segments stored in a s3 bucket and sort them by filename to create a VOD playlist. But then I can just use the metadata (last modified) for sorting. Therefore all good :)

**Comment by gabek** [comment]
Ok, hopefully that should clear that up, it just uses an incremented segment counter now, so a collision can't occur.  Let me know if you run into any issues with it.  Thanks again!

**Comment by zerodytrash** [comment]
wow that was quick! I will test tomorrow. 
Thank you!
